### PR TITLE
bpo-44166: Make IndexError messages for lists more informative

### PR DESCRIPTION
**Pull request by miguendes**
https://bugs.python.org/issue44166

### CODE REVIEW

File: Objects/listobject.c [hunk 1]
```diff
@@ -218,7 +218,20 @@ valid_index(Py_ssize_t i, Py_ssize_t limit)
...
+index_error(const char *msg, Py_ssize_t len, Py_ssize_t i)
+{
+    if (len == 0) {
+        PyErr_Format(PyExc_IndexError, "%s index out of range, got index %zd but the list is empty", msg, i);
+        return;
+    }
+    if (i < 0)
+        i -= len;
+    PyErr_Format(PyExc_IndexError,
+                 "%s index out of range, the len is %zd so index must be in -%zd..%zd, got %zd",
```
Comment (Fidget-Spinner) [inline Objects/listobject.c:231]: Wording:
```suggestion
                 "%s index out of range, the length is %zd so index must be between -%zd..%zd, got %zd",
```

File: Objects/listobject.c [hunk 2]
```diff
@@ -218,7 +218,20 @@ valid_index(Py_ssize_t i, Py_ssize_t limit)
...
-static PyObject *indexerr = NULL;
+static inline void
+index_error(const char *msg, Py_ssize_t len, Py_ssize_t i)
+{
+    if (len == 0) {
+        PyErr_Format(PyExc_IndexError, "%s index out of range, got index %zd but the list is empty", msg, i);
+        return;
+    }
+    if (i < 0)
+        i -= len;
```
Comment (Fidget-Spinner) [inline Objects/listobject.c:229]: formatting: PEP 7 needs brackets on new code
```suggestion
    if (i < 0) {
        i -= len;
    }
```

**Review by pablogsal — CHANGES_REQUESTED** [review]
This is a nice Idea. One thing I'm concerned about is that this removed the optimization for creating the string of index errors. As we are dynamically creating the strings now it means that code that raises index error but catches it later will be slower. 

We need to run some benchmarks to be sure that we are not going to introduce a big regression.

**Review by Fidget-Spinner — COMMENTED** [review]
I just have a few minor suggestions below. Apart from what Pablo said, the rest of the code LGTM.

**Comment by Fidget-Spinner** [comment]
BTW CI is failing because GitHub actions has some issues right now https://www.githubstatus.com/history. I doubt it's due to the code.

**Comment by miguendes** [comment]
> This is a nice Idea. One thing I'm concerned about is that this removed the optimization for creating the string of index errors. As we are dynamically creating the strings now it means that code that raises index error but catches it later will be slower.
> 
> We need to run some benchmarks to be sure that we are not going to introduce a big regression.

Thanks for your input @pablogsal !

I ran some benchmarks. And indeed this new change is slower, the optimised version is around 1.82x faster.

```bash
$ ./python bench_index_error.py -o new.json
.....................
index_error: Mean +- std dev: 676 ns +- 13 ns

$ ./python bench_index_error.py -o old.json
.....................
index_error: Mean +- std dev: 372 ns +- 5 ns

$ ./python -m pyperf compare_to new.json old.json
Mean +- std dev: [new] 676 ns +- 13 ns -> [old] 372 ns +- 5 ns: 1.82x faster
```

Here's the code I used to benchmark it: https://gist.github.com/miguendes/99d7d57c9d48bf121ce7a30b190009bb

I fully understand the performance hit and it's a shame we cannot make the user experience better for this use case :(. Especially because this would be aligned with the new improvements we've done for SyntaxErrors. Even though we can't compare them in terms of performance.

Out of curiosity, how much slower should it be to be considered a big regression? 

Also, is the EAFP style mentioned by @serhiy-storchaka so common that this change would be a no-go? 
I was assuming that even if we catch these exceptions, this kind of event wouldn't happen so frequently, but I might be wrong.

Is there anything I can do to make it faster?

**Comment by pablogsal** [comment]
> Out of curiosity, how much slower should it be to be considered a big regression?

It depends, but generally more than 4-5 % starts to raise concerns.

> Also, is the EAFP style mentioned by @serhiy-storchaka so common that this change would be a no-go?

Yep, is not the most common code, but is important enough to make this approach at least probably a no-go.

I have some ideas on how to implement this with no performance cost, but I need to finish something first before I can explain it in detail.

**Comment by rhettinger** [comment]
As stated on the tracker, we've decided to leave the code as-is.

Thank you for submitting a PR.

**Pull request closed without merging**